### PR TITLE
Flex query

### LINKED ISSUE #42: Flexible Query Part 1: Store
1. IDStore -> all resources indexed by UUID
2. TypeStore -> all resources indexed by type basically ResourceMap where the key is typeName
3. LabelStore -> all resources indexed by labelKey and value map[string]ResourceMap
4. FileStore -> all resources that will be persisted to disk

All the above stores will implement Store Interface and custom Query interface. For mutation operations CUD we will use a ResourceStore that will call the operation on all stores and each store will handle the mutation correctly. The ResourceStore will also implement all the custom query interfaces like IDQuery, LabelQuery and TypeQuery to and provides correct forwarding based on the type of the query.

### PR DESCRIPTION
**Pull request by shravyanandyala**
Fixes #42

### CODE REVIEW

File: typestore/typestore.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/project-safari/zebra"
+)
+
+type TypeStore struct {
+	lock      sync.RWMutex
```
Comment (rchamarthy) [inline typestore/typestore.go:10]: assuming the outer store is the only way to access these specific stores for read/write we can keep these lock free and implement safety mechanisms in the outer store that will coordinate. Lets think about it.

File: typestore/typestore.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+	ErrUpdateNoExist = errors.New("called update on a resource that does not exist")
+	ErrNotFound      = errors.New("resource not found in store")
+)
+
+// Return new query store pointer given resource map.
+func NewTypeStore(resources *zebra.ResourceMap) *TypeStore {
+	typestore := &TypeStore{
+		lock: sync.RWMutex{},
+		resources: func() *zebra.ResourceMap {
+			dest := zebra.NewResourceMap(nil)
```
Comment (rchamarthy) [inline typestore/typestore.go:19]: shouldnt this get the factory from a contract pov?

File: typestore/typestore.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+	return nil
+}
+
+// Return all resources in a ResourceMap.
+func (ts *TypeStore) Load() (*zebra.ResourceMap, error) {
+	ts.lock.RLock()
+	defer ts.lock.RUnlock()
+
+	resources := zebra.NewResourceMap(nil)
```
Comment (rchamarthy) [inline typestore/typestore.go:56]: this one should also get in the factory imo

File: typestore/typestore.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+// Should not be called without holding the write lock.
+func (ts *TypeStore) delete(res zebra.Resource) error {
+	ts.resources.Delete(res, res.GetType())
+
+	return nil
+}
+
+// Return all resources of given types in a ResourceMap.
+func (ts *TypeStore) Query(types []string) *zebra.ResourceMap {
+	retMap := zebra.NewResourceMap(ts.resources.GetFactory())
```
Comment (rchamarthy) [inline typestore/typestore.go]: if we are going to have lock in this store then this should have the read lock before reading the data.

File: typestore/typestore.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+	for _, t := range types {
+		list := zebra.NewResourceList(ts.resources.GetFactory())
+		zebra.CopyResourceList(list, ts.resources.Resources[t])
```
Comment (rchamarthy) [inline typestore/typestore.go:123]: assuming only the outer store will use this inner store, we should not copy the resources here, we should let the outer store do the query, apply all the filters and before doing the final return we can do the copy. That way we only copy the final result. This does mean that the internal stores cannot be accessed by anything else other the managing store.

File: typestore/typestore_test.go [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+	resMap := zebra.NewResourceMap(nil)
+	resMap.Add(new(network.VLANPool), "VLANPool")
+	resMap.Add(new(network.IPAddressPool), "IPAddressPool")
+
+	ts := typestore.NewTypeStore(resMap)
+	assert.NotNil(ts)
+
+	resources, err := ts.Load()
+	assert.Nil(err)
+	assert.True(len(resources.Resources) == 2)
```
Comment (rchamarthy) [inline typestore/typestore_test.go:57]: use assert.Equal

File: typestore/typestore_test.go [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+	assert.Nil(ts.Clear())
+
+	resources, err = ts.Load()
+	assert.Nil(err)
+	assert.True(len(resources.Resources) == 0)
```
Comment (rchamarthy) [inline typestore/typestore_test.go:63]: assert.Empty

File: labelstore/labelstore.go [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+type Query struct {
+	Op     Operator
+	Key    string
+	Values []string
+}
+
+type LabelStore struct {
+	lock      sync.RWMutex
+	factory   zebra.ResourceFactory
+	uuids     map[string]zebra.Resource
```
Comment (rchamarthy) [inline labelstore/labelstore.go:29]: why the UUID map?

File: labelstore/labelstore.go [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+// Return all resources in a ResourceMap where keys are labelName = labelVal.
+func (ls *LabelStore) Load() (*zebra.ResourceMap, error) {
+	ls.lock.RLock()
+	defer ls.lock.RUnlock()
+
+	retMap := zebra.NewResourceMap(ls.factory)
+
+	for label, valMap := range ls.resources {
```
Comment (rchamarthy) [inline labelstore/labelstore.go:94]: if you have the uuid based map you can make a copy from that is much simpler. why make the resource map manually and not use the Add method on the ResourceMap. Also I noticed the ResourceMap Add takes the resource and the type string, you could get the type string from the resource.GetType so why need extra param.

File: labelstore/labelstore.go [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+// Should not be called without holding the write lock.
+func (ls *LabelStore) delete(res zebra.Resource) error {
+	// If resource does not exist in store, just return without error
+	if _, err := ls.find(res.GetID()); err != nil {
+		return nil
+	}
+
+	for label, val := range res.GetLabels() {
+		if ls.resources[label] != nil {
+			ls.resources[label].Delete(res, val)
```
Comment (rchamarthy) [inline labelstore/labelstore.go:170]: if all the resources for this label are deleted we should delete the entry for the label itself. Other wise will keep hold of the bucket for this label for ever and indirectly leak memory. For example i keep create and delete labels your map will keep increasing in memory. Let me know if we need discuss this.

File: labelstore/labelstore.go [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+		}
+	}
+
+	return nil
+}
+
+// Return all resources of given label - label value pairs in a ResourceMap.
+func (ls *LabelStore) Query(query Query) *zebra.ResourceMap {
+	switch query.Op {
```
Comment (rchamarthy) [inline labelstore/labelstore.go:179]: read lock the same way if we keep the lock

File: labelstore/labelstore.go [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+	return ls.delete(res)
+}
+
+// Should not be called without holding the write lock.
+func (ls *LabelStore) delete(res zebra.Resource) error {
+	// If resource does not exist in store, just return without error
+	if _, err := ls.find(res.GetID()); err != nil {
+		return nil
+	}
+
```
Comment (rchamarthy) [inline labelstore/labelstore.go:167]: how about deleting from the uuids cache as well?

File: resourcestore/resourcestore.go [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+// Should not be called without holding the write lock.
+func (rs *ResourceStore) delete(res zebra.Resource) error {
+	delete(rs.resources, res.GetID())
+
+	return nil
+}
+
+// Return all resources of given UUIDs in a ResourceMap.
+func (rs *ResourceStore) Query(ids []string) *zebra.ResourceMap {
+	retMap := zebra.NewResourceMap(rs.factory)
```
Comment (rchamarthy) [inline resourcestore/resourcestore.go]: read lock if we keep it

File: store/store.go [hunk 14]
```diff
@@ -1,19 +1,164 @@
...
+	rs.fs = nil
+	rs.ids = nil
+	rs.ls = nil
+	rs.ts = nil
+
+	return nil
+}
+
+func (rs *ResourceStore) Clear() error {
+	_ = rs.fs.Clear()
```
Comment (rchamarthy) [inline store/store.go:60]: you should return the error if any of these substores actually return an error

File: store/store.go [hunk 15]
```diff
@@ -1,19 +1,164 @@
...
+
+	rs.ids = idstore.NewIDStore(resources)
+	rs.ls = labelstore.NewLabelStore(resources)
+	rs.ts = typestore.NewTypeStore(resources)
+
+	return nil
+}
+
+func (rs *ResourceStore) Wipe() error {
+	rs.fs = nil
```
Comment (rchamarthy) [inline store/store.go:51]: you need to lock

File: store/store.go [hunk 16]
```diff
@@ -1,19 +1,164 @@
...
+
+func (rs *ResourceStore) Create(res zebra.Resource) error {
+	if res == nil || res.Validate(context.Background()) != nil {
+		return zebra.ErrInvalidResource
+	}
+
+	rs.lock.Lock()
+	defer rs.lock.Unlock()
+
+	err := rs.fs.Create(res)
```
Comment (rchamarthy) [inline store/store.go:81]: a little problem with transaction sematics. We need to implement. Lets finish this PR and create an issue for it. We need to implement a simple 2phase commit.

**Review by rchamarthy — CHANGES_REQUESTED** [review]

File: labelstore/labelstore.go (hunk 10 above)
Comment (shravyanandyala) [reply labelstore/labelstore.go:170]: nope I understand, oversight on my part.

**Review by rchamarthy — APPROVED** [review]

**Pull request merged**